File: Day08/p1.py
def parse_instructions(instructions):
    result = []
    for instruction in instructions:
        ins, amount = instruction.split(' ')
        amount = int(amount)
        result.append([ins, amount, 0])
    return result


def execute(instructions):
    total = 0
    i = 0
    while instructions[i][0] is not None:
        instruction = instructions[i]
        if instruction[2] > 0:
            return total
        instruction[2] += 1

        action = instruction[0]
        if action == "acc":
            total += instruction[1]
        elif action == "jmp":
            i += instruction[1] - 1
        i += 1
    raise Exception("Code did not loop!")


with open("code.txt", "r") as file:
    instructions = file.read().splitlines()

instructions = parse_instructions(instructions)
# Add None to the end to serve as a break
instructions.append([None, 0, 0])

value = execute(instructions)
print(value)

exit(0)

File: Day08/p1.hs
module P1 where

import qualified Data.Set as Set  

data Instruction a = Nop a | Acc a | Jmp a deriving (Show)

main :: IO ()
main = do
    ops <- parseFile "code.txt"
    let acc = followInstructions ops 0 0 Set.empty 
    putStrLn $ "Accumulator finished at " ++ show acc


parseFile :: (Read a, Num a) => FilePath -> IO [Instruction a]
parseFile filename = do
    contents <- readFile filename
    return (map (parseInstruction . words) (lines . filter (/='+') $ contents))


parseInstruction :: (Read a, Num a) => [String] -> Instruction a
parseInstruction [op, sval]
    | op == "nop" = Nop val 
    | op == "acc" = Acc val
    | op == "jmp" = Jmp val 
    where val = read sval

followInstructions :: (Ord a, Integral a) => [Instruction a] -> a -> a -> Set.Set a -> a
followInstructions instructions currentIns acc visited
    | currentIns >= fromIntegral (length instructions) = acc
    | currentIns `Set.member` visited = acc
    | otherwise = case ins of 
        Nop x -> followInstructions instructions (currentIns + 1) acc updatedVisited
        Acc x -> followInstructions instructions (currentIns + 1) (acc + x) updatedVisited
        Jmp x -> followInstructions instructions (currentIns + x) acc updatedVisited
    where ins = instructions !! fromIntegral currentIns
          updatedVisited = currentIns `Set.insert` visited

